src/util/indexing.jl

```diff
@@ -0,0 +1,249 @@
+
+
+"""
+Construct an Indexer that maps a Vector of Symbols (names), to a set of
+values of type Int, where Int is the default integer encoding on you machine
+(usually Int32 or Int64).
+
+Provide this constructor a vector of Symbols as names, and it will construct
+an index such that names[1] -> 1, names[2] -> 2, ...,
+names[length(names)] -> length(names).
+"""
+function Indexer(names::Vector{Symbol})
+    return Indexer(names, Int)
+end
+
+
+"""
+Construct an Indexer that maps a Vector of Strings (names), to a set of
+values of type Int, where Int is the default integer encoding on you machine
+(usually Int32 or Int64).
+
+Provide this constructor a vector of Strings as names, and it will construct
+an index such that names[1] -> 1, names[2] -> 2, ...,
+names[length(names)] -> length(names).
+"""
+function Indexer{S <: AbstractString}(names::Vector{S})
+    return Indexer(convert(Vector{Symbol}, names), Int)
+end
+
+
+"Make sure all vectors in a symbol are unique."
+function make_unique(names::Vector{Symbol})
+    seen = Set{Symbol}()
+    names = copy(names)
+    dups = Int[]
+    for (i, name) in enumerate(names)
+        in(name, seen) ? push!(dups, i) : push!(seen, name)
+    end
+    for i in dups
+        nm = names[i]
+        k = 1
+        while true
+            newnm = symbol("$(nm)_$k")
+            if !in(newnm, seen)
+                warn("Trying to build an index with duplicate names. Changing $(names[i]) to $(newnm)")
+                names[i] = newnm
+                push!(seen, newnm)
+                break
+            end
+            k += 1
+        end
+    end
+    return names
+end
+
+
+Base.length(x::Indexer) = length(x.names)
+Base.names(x::Indexer) = copy(x.names)
+_names(x::Indexer) = x.names
+Base.copy(x::Indexer) = Indexer(copy(x.lookup), copy(x.names))
+Base.deepcopy(x::Indexer) = copy(x) # all eltypes immutable
+Base.isequal(x::Indexer, y::Indexer) = isequal(x.lookup, y.lookup) && isequal(x.names, y.names)
+Base.(:(==))(x::Indexer, y::Indexer) = isequal(x, y)
+Base.haskey(x::Indexer, key::Symbol) = haskey(x.lookup, key)
+Base.haskey{S <: AbstractString}(x::Indexer, key::S) = haskey(x, convert(Symbol, key))
+Base.haskey(x::Indexer, key::Real) = 1 <= key <= length(x.names)
+Base.keys(x::Indexer) = names(x)
+
+
+"Completely replace the names in an Indexer with a new Vector of Symbols."
+function names!(x::Indexer, names::Vector{Symbol})
+    if length(names) != length(x)
+        throw(ArgumentError("Length of new names doesn't match length of Index."))
+    end
+    names = make_unique(names)
+    if length(x.names) != length(x.lookup)
+        throw(ArgumentError("Index corrupted by duplicate symbols in names."))
+    end
+    rename!(x, x.names, names)
+    return x
+end
+
+
+"Completely replace the names in an Indexer with a new Vector of Strings."
+function names!{S <: AbstractString}(x::Indexer, names::Vector{S})
+    names!(x, convert(Vector{Symbol}, names))
+end
+
+
+function rename!(x::Indexer, names)
```

Comment:
Is it an issue that all the `rename` functions only work with `Symbol`? I'm going to assume not, but other operations are defined for both